### DISCUSSION TITLE
Upload and Download Progress

### DISCUSSION THREAD

**Discussion by vspr14** [pocketbase/pocketbase#1567] [Q&A]
Hi! I am using the dart SDK for file storage and I was wondering if there was any way to get the amount of bytes currently downloaded essentially to create a upload progress indicator. Currently I've implemented a download progress indicator since the Uri is available. For the upload, since a record is being created and file handling is not done explicitly, is there a possibility of adding the ability to get a progress stream?

**Comment by ganigeorgiev** [comment]
This was recently requested, but I'm not sure if there is a reliable cross-platform solution (note that the Dart SDK is not only for Flutter and it is intended to be used as a generic Dart lib).

It depends on a lot of factors, including how the `MultipartFile` instance was created/loaded (eg. `fromBytes`, `fromFile`, etc.).
If you are targeting only web, the browser's (`dart:html`) XHR provides an upload progress event that eventually could be used.
You can also check dart-lang/sdk#35427.

**Reply by vspr14** [reply]
I am not entirely sure if this even makes sense, but is there a way we can make a request to the pocketbase instance using the http module instead of using the create record method? There seem to be a few implementations of using a MultiPartRequest() function to achieve the same but this cant be used with the sdk.

**Reply by ganigeorgiev** [reply]
The SDK is built on top of `dart/http` but at the moment we don't expose the json or multipart request objects (I'll consider it in the future).

You can use directly the dart http client and send a request to the Record Create API - https://pocketbase.io/docs/api-records/ if you want to create your own custom handling.

**Reply by vspr14** [reply]
im not sure how to create my own handling from the link given.. it seems to go to the original pocketbase dart sdk docs.

**Reply by ganigeorgiev** [reply]
The above is a link to the HTTP api used by the SDK. Click on the "Create record" accordion to explore it (it contains both sdk examples and api details like required body parameters, request url, etc.).

**Reply by vspr14** [reply]
ah! so i can make a direct HTTP request using the  "/api/collections/collectionIdOrName/records" endpoint!

**Comment by vspr14** [comment]
I tried creating a new record using the API but every time i try to upload a file, it returns a status 400 but if i upload without the file, it seems to create a new record using the following code:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/63847431/212002415-85802c98-a1a6-412e-aa50-e334f28cf784.png">

**Reply by ganigeorgiev** [reply]
Are you sure that the request satisfy the Create API rule of your collection?

You can check the detailed error by inspecting the error response.

For a more detailed error message you can try starting PocketBase with `--debug` and/or check the request log details from "Admin UI > Logs".

It could be also something specific to `Dio`, I don't have much experience with that http client library (_the multipart field name doesn't seem to match with what you are sending in the fromMap_).
